qa/tasks/cephfs/test_client_recovery.py

```diff
@@ -629,3 +630,73 @@ def test_config_session_timeout(self):
         self.assert_session_count(1)
 
         self.mount_a.kill_cleanup()
+
+    def test_reconnect_after_blacklisted(self):
+        """
+        Test reconnect after blacklisted.
+        - writing to a fd that was opened before blacklist should return -EBADF
+        - reading/writing to a file with lost file locks should return -EIO
+        - readonly fd should continue to work
+        """
+
+        if not isinstance(self.mount_a, KernelMount):
+            self.skipTest("Not implemented in FUSE client yet")
+
+        self.mount_a.umount_wait()
+        self.mount_a.mount(mount_opts=['recover_session=clean'])
+        self.mount_a.wait_until_mounted()
+
+        path = os.path.join(self.mount_a.mountpoint, 'testfile_reconnect_after_blacklisted')
+        pyscript = dedent("""
+            import os
+            import sys
+            import fcntl
+            import errno
+            import time
+
+            fd1 = os.open("{path}.1", os.O_RDWR | os.O_CREAT, 0666)
+            fd2 = os.open("{path}.1", os.O_RDONLY)
+            fd3 = os.open("{path}.2", os.O_RDWR | os.O_CREAT, 0666)
+            fd4 = os.open("{path}.2", os.O_RDONLY)
+
+            os.write(fd1, 'content')
+            os.read(fd2, 1);
+
+            os.write(fd3, 'content')
+            os.read(fd4, 1);
+            fcntl.flock(fd4, fcntl.LOCK_SH | fcntl.LOCK_NB)
+
+            sys.stdout.write('\n')
+            sys.stdin.readline()
+
+            time.sleep(10);
+
+            try:
+                os.write(fd1, 'content')
+            except OSError as e:
+                if e.errno != errno.EBADF:
+                    raise
+            else:
+                raise RuntimeError("write() failed to raise error")
+
+            os.read(fd2, 1);
+
+            try:
+                os.read(fd4, 1)
+            except OSError as e:
+                if e.errno != errno.EIO:
+                    raise
+            else:
+                raise RuntimeError("read() failed to raise error")
+            """).format(path=path)
+        rproc = self.mount_a._run_python(pyscript)
+
+        rproc.stdout.readline()
+
+        mount_a_client_id = self.mount_a.get_global_id()
+        self.fs.mds_asok(['session', 'evict', "%s" % mount_a_client_id])
```

Comment:
Silly q: does it matter that we aren't actually blacklisting the client here? Might the behavior be different if the first attempt to reconnect after the session eviction were to fail due to blacklisting?

Reply:
It does. dirty data is not dropped if not blacklisted. writing to old fd does not return error